### ISSUE TITLE
Full Comfy Support Lacking

### ISSUE THREAD

**Issue by gseth** [NVlabs/Sana#176]
Hi,

I have extensively tried both the Sana Versions in COmfy UI and Your Inference. There is a significant difference, and it is **due to the lack of Flow-DPM-Solver support in Comfy UI.** My Comfy is on a different PyTorch and CUDA; hence, the same seed provides a minor variation in the image.

I have tried extensive Detailed LLM Prompts in comfy which are curated for the Gemma 2 Model being used, the image quality improves but its no way near a simple prompt in your original inference.

I have even tried double sampling in Comfy UI with latent upscale with some interesting results.

1024 - Latent Upscale by 2x doesn't work. The second Pass is worse than the first pass.
2048 - Works. The second pass fixes the eyes faces etc which are a major issue.
4096 - not enough VRAM to try at the moment on a 4090.

Besides Flow-DPM-Solver being missing, You need a Sana Guidance Node. This I am not sure cause I did not study the architecture in that detail. However I can confirm that Lowering and Increasing Guidance works differently in Comfy and your inference. I am not sure if this is due to a lack of a dedicated guidance node or the missing Sampler.

Along with the Guidance Node, The PAG build-in node in Comfy has zero effect on the image. Whereas in inference, the PAG of 1-4 is drastic.

Steps is another thing where in Comfy the recommended 28-50 steps is basically so minor, there is no reason to go to 50. However, in the inference code it makes a drastic difference between the steps range.

Please note, I have tested from Human to Artistic and various other examples of my own and some using the official prompts given on the main page.  With only positive and then both positive and negative. I could improve image quality by changing some techniques in comfy but at the current state the inference is far far superior.

Thank you for your work on Sana, and please let me know if you’d like any additional details from my testing process!

**Comment by lawrence-cj** [comment]
Thanks so much for your effort on testing Sana. I think the main reason for the difference comes from the Flow-DPM-Solver from our repo and Flow-Euler in the ComfyUI node. Implementing Flow-DPM-Solver in ComfyUI is not the highest priority stuff in our progress actually.

Beside the results above, can you share some insight on what guidance, auxiliary or techs can make Sana perform best in ComfyUI in your experience?

**Comment by gseth** [comment]
@lawrence-cj Sorry it took some time. We were doing research, and that took longer than expected, but here is the complete analysis.

Comfy UI: Observations
1. Best setting is steps 20-28. Use of LLM is recommended. The use of negative prompts is recommended. CFG 2.-7 depending on the style of image. 
2. You use SANA Clip Node for Positive and Negative in your sample workflow. And for the 4k Workflow Sana, one is only used for positive and Gemma for Negative. As per our testing, Sana should be used only for Positive and Gemma for Negative in all workflows and models. When Sana only is used for positive and negative We cannot go beyond CFG 2. The same behavior is seen when Gemma is used for both positive and negative.
3. The different samplers in Comfy behavior, like Samplers+VAE (SD1.5). Some samplers don't do anything but drastically enhance the brightness and color grading of the image. Like DPM_2 (not sure which ones did that). This behavior was typical with VAE in SD 1.5, where using the wrong VAE would give an undersaturated, darker image. This is just an observation; I don't know what this behavior leads to. Since you have implemented it maybe it helps.

ComfyUI Integration Advise:
1. A dedicated Node for flow sampler. Inference has flow sampler at different values for different models
2. Flow DPM Solver is a must. Either include that or make a Cusomt Sana Sampler which user the same Flow DPM Solver Sampler/Scheduler. A custom Sana Sampler node is much more appropriate in this case. Half the things in the K sampler won't work, like karras.
3. PAG attention injection at layer 8. A dedicated node which injects PAG scale 1-4 as observed in inference at layer 8. You can do an attention overide I have not studied the attention architecture in detail or how its injected. What I am saying is PAG injection is crucial. In inference testing the PAG in combination of CFG makes a hell lot of difference. If I want to go with landscapes I reduce CFG to as low as 2-3, and I get beautiful images. For specific game assets I would increase them.
4. Sana ControlNet node. Your inference controlnet is very very impressive. I was shocked to see results with a simple HED implementation. To get similar results in Flux, I have to merge and do a very complicated workflow. 

Blackwell Architexture Issues.
1. For Blackwell we need the 12.8 build for linux / wsl. Your requirements overides everything. To solve this, I first installed all your requirements step by step. Last being the Sana install. Then I understalled torch and installed nightly one 2.7. This stopped giving kernel errors on blackwell architecture. But xformers now was an issue. So the simplest solution was to uninstall xformers since the latest 0.29 version doesnt' work with pytorch 2.7 and cuda 12.8.
2. Since xformers was not there, I later learned that in inference you use xformers for the Control Net and PyTorch’s scaled dot product attention as a fallback. Now since I don't have xformers, your fallback code was broken. So I fixed it (I am not a coder, so excuse if there are mistakes.)

Fall Back code fix for blackwell architecture (till xformers is compatible):

sana_blocks.py

```
            q, k, v = q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2)
            if mask is not None:
                if isinstance(mask, list):
                    mask = torch.tensor(mask, device=x.device)
                # Cast mask to the same dtype as q (e.g. bfloat16) instead of default float32.
                mask = mask.to(q.dtype)
                # If mask is 1D, assume it has shape [B] and expand it to [B, key_length].
                if mask.ndim == 1:
                    key_len = k.shape[2]
                    mask = mask.unsqueeze(1).repeat(1, key_len)
                # Now mask is 2D ([B, key_length]); convert to a 4D tensor.
                mask = (1 - mask) * -10000.0
                mask = mask[:, None, None].expand(mask.size(0), self.num_heads, q.shape[2], mask.size(1))
            x = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0, is_causal=False)
            x = x.transpose(1, 2)
```

In both places of the file. After this I could do Controlnet on 5090.

I have educated the end/business users via a video tutorial covering all these points: https://youtu.be/npw9qc8mOmQ

If you require any help, please reach out. Happy to contribute in whatever way I can / possible.
